hashicorp/vault
# Fixes redirects to KV engine when secret is a directory

Thank you @kiannaquach for a fix that I found on your quick actions card.

**To reproduce:**
_not on this branch just on your most recent binary 1.15.0+_
1. Create a kv engine with a nested secret, something like `kv-test` with secret `beep/bop/boop`.
2. Make a policy for a user bob that allows list access only to the beep/bop folder.
```
path "kv-test/metadata/beep/bop" {
  capabilities = ["list"]
}
path "kv-test/data/beep/bop/boop" {
  capabilities = ["read", "create", "update"]
}
```
3. Login as root and navigate to the 'bop' folder. ex: `http://localhost:8200/ui/vault/secrets/kv-test/kv/list/beep/bop/`
4. Copy the URL, and in an incognito window login paste that URL **_without_** the kv in the URL **and login as bob**. ex: `http://localhost:8200/ui/vault/secrets/kv-test/list/beep/bop/`
5. The URL redirect from the old list view does not work. If you add back in the kv to the URL and login in as bob again, you'll see the view you should be redirected to. The expectation is that the URL should redirect from the old routing structure to the new one.
This is the incorrect view (note: the button is fixed in another pr).
<img width="1503" alt="image" src="https://github.com/hashicorp/vault/assets/6618863/eebbd69b-9010-4152-a346-243ccdeac949">

**To test this fix:**

1. Do the same steps as above but in local dev with this branch checked out. On step 4, you should be redirected to the directory view of beep/bop.

<img width="1113" alt="image" src="https://github.com/hashicorp/vault/assets/6618863/31f64a31-1188-4d5d-b2de-135d9faf4ef5">
